components/remote_settings/src/lib.rs

```diff
@@ -50,6 +60,80 @@ impl RemoteSettings {
     }
 }
 
+pub struct RemoteSettingsBuilder {
+    inner: Mutex<RemoteSettingsBuilderInner>,
+}
+
+#[derive(Clone, Default)]
+struct RemoteSettingsBuilderInner {
+    server: Option<RemoteSettingsServer>,
+    bucket_name: Option<String>,
+    collection_name: Option<String>,
+    cache: Option<Arc<dyn RemoteSettingsCache>>,
+}
+
+impl Default for RemoteSettingsBuilder {
+    fn default() -> Self {
+        Self {
+            inner: Mutex::new(RemoteSettingsBuilderInner::default()),
+        }
+    }
+}
+
+impl RemoteSettingsBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn collection_name(self: Arc<Self>, collection_name: String) -> Arc<Self> {
```

Comment:
I wonder why we return `Arc` instead of `Self` directly?

Typically in a Builder Pattern, one might expect a `Self`.

```rust
  pub fn collection_name(mut self, collection_name: String) -> Result<Self, RemoteSettingsError> {
        let mut inner = self.inner.lock().map_err(|_| RemoteSettingsError::LockError)?;
        inner.collection_name = Some(collection_name);
        Ok(self)
    }
```

But not sure how devs might expect to use it, so just throwing it out there.

Reply:
It's a UniFFI thing.  Right now the only way to express this is an an interface and interfaces always need to be wrapped in an Arc, since that's how objects work in all the foreign languages.

I think we could avoid this if we supported methods on records/enums.  That would also remove the need for the inner mutex.